include/misc/mutex.h

```diff
@@ -0,0 +1,149 @@
+ */
+
+#ifdef CONFIG_USERSPACE
+#include <atomic.h>
+#include <zephyr/types.h>
+
+struct sys_mutex {
+	/* Currently unused, but will be used to store state for fast mutexes
+	 * that can be locked/unlocked with atomic ops if there is no
+	 * contention
+	 */
+	atomic_t val;
+};
+
+#define SYS_MUTEX_DEFINE(name) \
+	struct sys_mutex name
+
+/**
+ * @brief Initialize a mutex.
+ *
+ * This routine initializes a mutex object, prior to its first use.
+ *
+ * Upon completion, the mutex is available and does not have an owner.
+ *
+ * This routine is only necessary to call when userspace is disabled
+ * and the mutex was not created with SYS_MUTEX_DEFINE().
+ *
+ * @param mutex Address of the mutex.
+ *
+ * @return N/A
+ */
+static inline void sys_mutex_init(struct sys_mutex *mutex)
+{
+	ARG_UNUSED(mutex);
+
+	/* Nothing to do, kernel-side data structures are initialized at
+	 * boot
+	 */
+}
+
+__syscall int z_sys_mutex_kernel_lock(struct sys_mutex *mutex, s32_t timeout);
+
+__syscall int z_sys_mutex_kernel_unlock(struct sys_mutex *mutex);
+
+/**
+ * @brief Lock a mutex.
+ *
+ * This routine locks @a mutex. If the mutex is locked by another thread,
+ * the calling thread waits until the mutex becomes available or until
+ * a timeout occurs.
+ *
+ * A thread is permitted to lock a mutex it has already locked. The operation
+ * completes immediately and the lock count is increased by 1.
+ *
+ * @param mutex Address of the mutex, which may reside in user memory
+ * @param timeout Waiting period to lock the mutex (in milliseconds),
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval 0 Mutex locked.
+ * @retval -EBUSY Returned without waiting.
+ * @retval -EAGAIN Waiting period timed out.
+ * @retval -EACCESS Caller has no access to provided mutex address
+ * @retval -EINVAL Provided mutex not recognized by the kernel
+ */
+static inline int sys_mutex_lock(struct sys_mutex *mutex, s32_t timeout)
+{
+	/* For now, make the syscall unconditionally */
+	return z_sys_mutex_kernel_lock(mutex, timeout);
+}
+
+/**
+ * @brief Unlock a mutex.
+ *
+ * This routine unlocks @a mutex. The mutex must already be locked by the
+ * calling thread.
+ *
+ * The mutex cannot be claimed by another thread until it has been unlocked by
+ * the calling thread as many times as it was previously locked by that
+ * thread.
+ *
+ * @param mutex Address of the mutex, which may reside in user memory
+ * @retval -EACCESS Caller has no access to provided mutex address
+ * @retval -EINVAL Provided mutex not recognized by the kernel or mutex wasn't
+ *                 locked
+ * @retval -EPERM Caller does not own the mutex
+ */
+static inline int sys_mutex_unlock(struct sys_mutex *mutex)
+{
+	/* For now, make the syscall unconditionally */
+	return z_sys_mutex_kernel_unlock(mutex);
+}
+
+#include <syscalls/mutex.h>
+
+#else
+#include <kernel.h>
+#include <kernel_structs.h>
+
+struct sys_mutex {
+	struct k_mutex kernel_mutex;
```

Comment:
So... this pointer lives in userspace and can be scribbled on.  But the pointer is to a k_mutex object in kernel space, so we can test that for validity via normal means, so we can rely on it being a k_mutex pointer.  But... the user could have set that pointer to any value.  We can't rely on it being the **correct** k_mutex pointer.

Can that break anything?  I mean, it's possible for an attacker to lock or unlock any k_mutex to which they have access.  And given that those calls can drop or boost the priority of the process.... is that OK?  I can't think of a way to use that as an attack to do anything more than DoS the local memory space, but at the same time I'm not seeing a way to prove to myself that worse exploits aren't possible...

Basically: do we want to put this mapping of sys_mutex to k_mutex into kernel space?  I think we probably do.

Reply:
@andyross it is.
This definition is only when userspace is disabled.
When userspace is enabled, sys_mutex is just an atomic_t and the underlying k_mutex is generated by gen_kobject_list.py and stored in the gperf table.